File: LeetCode/LeetCode/insert_interval.cc
#include<iostream>
#include<vector>

using std::vector;
using std::max;
using std::min;
struct Interval {
    int start;
    int end;
    Interval() : start(0), end(0) {}
    Interval(int s, int e): start(s), end(e) {}
};

class Solution {
    public:
        vector<Interval> insert(vector<Interval>& intervals, Interval newInterval) {
            size_t i = 0;
            vector<Interval> result;

            while(i < intervals.size() && newInterval.start > intervals[i].end) {
                result.emplace_back(intervals[i++]);
            }

            while(i < intervals.size() && newInterval.end >= intervals[i].start) {
                newInterval = {min(newInterval.start, intervals[i].start), max(newInterval.end, intervals[i].end)};
                ++i;
            }
            result.emplace_back(newInterval);

            result.insert(result.end(), intervals.cbegin() + i, intervals.cend());
            return result;
        }
};

std::ostream& operator<<(std::ostream& o, std::vector<Interval> const& v) 
{
    for(auto i = v.begin(), end = v.end(); i != end; ++i) {
        o << "{"<< i->start<<"-"<< i->end <<"} ";
    }
    return o;
}

int main()
{
    Solution s;
    Interval newin = {3, 6};
    vector<Interval> vs = {
        {1,2},
        {4,6},
        {7,9}
    };
    std::cout<<s.insert(vs, newin)<<std::endl;
}

File: LeetCode/LeetCode/insert_interval.go
package main

import "fmt"

type Interval struct {
	Start int
	End int
}

func insert(a []Interval, newItv Interval) []Interval {
	lenA := len(a)

	if lenA == 0 {
		return []Interval{newItv}
	}

	if newItv.End < newItv.Start {
		return append(a, newItv)
	}

	res := make([]Interval, 0, len(a))
	for i := range a {
		if isOverlap(a[i], newItv) {
			newItv = merge(a[i], newItv)

			if i == lenA - 1 {
				res = append(res, newItv)
			}

			continue
		}

		if a[i].End < newItv.Start {
			res = append(res, a[i])
			continue
		}

		if newItv.End < a[i].Start {
			res = append(res, newItv)
			res = append(res, a[i:]...)
			break
		}
	}

	return res
}

func isOverlap(a, b Interval) bool {
	return (a.Start <= b.Start && b.Start <= a.End)
}

func merge(a, b Interval) Interval {
	return Interval{
		Start : min(a.Start, b.Start),
		End:	max(a.End, b.End),
	}
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func max(a, b int) int {
	if a>b {
		return a
	}
	return b
}


func main() {
	ni := Interval{3, 6}

	an := []Interval{
		{1,2},
		{4,6},
		{8,9},
	}

	fmt.Println(insert(an, ni))
}
